backends/p4tools/common/core/z3_solver.cpp

```diff
@@ -299,15 +299,11 @@ const IR::Literal *Z3Solver::toLiteral(const z3::expr &e, const IR::Type *type)
 }
 
 void Z3Solver::toJSON(JSONGenerator &json) const {
-    json << json.indent << "{\n";
-    json.indent++;
-    json << json.indent << "\"checkpoints\" : " << checkpoints;
-    json << ",\n";
-    json << json.indent << "\"declarations\" : " << declaredVarsById;
-    json << ",\n";
-    json << json.indent << "\"assertions\" : " << p4Assertions;
-    json.indent--;
-    json << json.indent << "}\n";
+    auto state = json.begin_object();
```

Comment:
Just loudly thinking: (not for this PR) we might do some scope-level RAII objects that would do `end_object` automagically. Or make `begin_object` to accept a lambda. So, `end_object` would be called automatically. I would probably prefer RAII as it might have nesting neater.

Reply:
The point of RAII is to recover from exceptions cleanly.  However, an exception when generating json would likely mean the entire json was invalid, so there's not much point in trying to fix up the superficial syntax, so there's not really anything useful for the RAII to recover/clean up.

We could add it (or a lambda) as an alternate syntax, however.

Reply:
Well, why cannot `state` call `end_object` in its destructor? We'd end with something like this:
```c++
{
  auto objectScope = json.begin_object_scope();
  object.emit("checkpoints", checkpoints);
  ...
  object.emit("assertions", p4Assertions);
}
```
Here `objectScope` would hold reference to `json`, so it could close the scope properly on destruction.

or, alternatively, with function object:
```c++
{
  json.object([](auto &state) {
    state.emit("checkpoints", checkpoints);
    ...
    state.emit("assertions", p4Assertions);
 });
}

Note sure which variant I like better.

Reply:
Ok, I made the 'state' value returned by `begin_object`/`_vector` automatically call `end_object`/`_vector` as appropriate when it goes out of scope, if it hasn't been called yet.  So now you can have an explicit `end` call, or you can just leave it off.

Since all these functions are inline in JSONGenerator.h, the overhead should be minimal -- it should be able to optimize away the redundant state changes and checks, and the result should be just a series of ostream outputs.